### DISCUSSION TITLE
CubeSat Development Process From End to End

### DISCUSSION THREAD

**Discussion by dtriska** [nasa/fprime#2286] [Q&A] [answered]
Hi everyone!
I firstly want to thank you all for building such a great community and providing this software which I know will be invaluable to my team and I. I am a part of an FSW team for my university CubeSat Project and I was hoping to get advice from the community as we embark on the development process.

 We are looking to follow the steps laid out here https://nasa.github.io/fprime/UsersGuide/best/development-practice.html however I was hoping the community could help delve a bit deeper into what to expect at each of these phases of the journey!
 
 Currently, we are still in the high-level design phase. We have laid out our requirements, and are currently building state diagrams for each of the states our CubeSat will be in. Furthermore, we are generating component breakdowns, which are meant to serve as a guide to our team detailing the interfaces and hardware capabilities of each our components. We plan to reference this with the F´ framework to see what functionality can be inherited for free.
 
 Next will be the Setup Deployment. We were wondering if you would recommend an in-tree deployment or a standalone deployment. We were leaning towards standalone, but are wanting feedback!
 
 The next step is where we aren't entirely sure as to the best course of action. We already have each member of our team assigned to a component. However, we were wondering when it comes to implementing this step should we adopt something like the Agile Methodology? In this case, we would set a goal and look to assemble the necessary components and have them work together to achieve that one specific goal/requirement. For example, take an image captured and send it to the GCS.  Or are we overthinking the process and underestimating how much F' can help us? Would we be able to make each component have all the necessary capabilities and then make the necessary processes work? What should the process look like at this point?
 
 Thank you for taking the time to read this all! I know it was a lot, but I am just trying to do everything I can to best lead my team! This is the most exciting thing that I have ever partaken in, and I want to pour everything I have into it, and hopefully not lead my team down the wrong path!
 
 One last thing, if anyone has any experience with the Kryten M3 OBC I would love to hear any advice you may have for me! And if you have any recommended CubeSat repos that are functional and could help guide us as an example that would be great as well! Also, any recommended high-level designs so we can compare them with ours, compare our requirements/diagrams, and ensure we fully understand the design of our system (I've been told to not skimp on the design phase) ! Thanks again!

**Comment by thomas-bc** [comment]
Hi @dtriska, glad to have you part of the community!
I'll make this a numbered list for convenience:

1) Would you be able to share which university your project is part of? We're always eager to build relationships, or simply know who's using our software!
2) Yes, go with the standalone deployment. This is now the only recommended way of doing things, and we have utilities to help you get set up really easily. If you follow any of the [tutorials](https://nasa.github.io/fprime/Tutorials/README.html) or the [install guide](https://nasa.github.io/fprime/INSTALL.html), it should walk you through how to do so.
3) Agile methodology can definitely help in some cases but it's also about figuring out what works the best for your team. As for what can F´ help with, I would say "a lot". The [SystemReference](https://github.com/fprime-community/fprime-system-reference) is an example application that we maintain, it features a Camera component so you can get inspiration from there if you want. Following the [LedBlinker tutorial](https://fprime-community.github.io/fprime-workshop-led-blinker/) can also be helpful to understand one recommended way to approach the development process, i.e. Component design -> implementation -> unit test -> integration test.
4) [SystemReference](https://github.com/fprime-community/fprime-system-reference) is a good example of a full F´ application that we maintain. Also, the University of Hawaii has been developing https://github.com/fprime-community/fprime-artemis-cubesat/tree/dev, which has 2 OBC, a RPI and a Teensy (so 2 F´ deployments interacting with each other).

**Reply by dtriska** [reply]
Hi @thomas-bc , and thank you for your reply!

I am on the CubeSTEP team at Cal Poly Pomona!

The information you have provided us with will be invaluable to us, and I know it will help steer us in the right direction moving forward!
(And thank you for giving specific and thoughtful feedback! I really can't thank you enough!) I particularly like the approach you suggested of Component design -> implementation -> unit test -> integration test. As you said, I think it'll be important to see which process works best for our specific team, but we will look to explore all approaches! This will definitely be at the top of the list!

I do have one other quick question I'm hoping you could answer for me! I've read through various papers on FSW design for CubeSat systems, and they all mention the immense importance of the design phase of the project, but they oftentimes lack detailed specifics of what the team actually does in this phase. I was curious if based on what I described in my initial post, do you believe we are taking proper strides in the high-level design. And if you had any suggestions for this paramount phase of the process.

Thank you again! (And thank you for sending the helpful reference links!)

**Reply by thomas-bc** [reply, marked as answer]
@dtriska maybe some of these slide decks can help answer some of these high-level questions you have: https://github.com/fprime-community/fprime-course-materials/tree/master/Flight_Software_Workshop-October_2023
Specifically the 03_FSW_Architecture and 13_FSW_Dev_Process
